### PR TITLE
Limit the count of exceptions inside a MetricPublisherPublishMessage

### PR DESCRIPTION
**Pull request by maosongfu**
#1475

### CODE REVIEW

File: heron/common/src/java/com/twitter/heron/common/config/SystemConfig.java [hunk 1]
```diff
@@ -540,6 +546,12 @@ public int getMetricsMgrNetworkOptionsSocketSendBufferSizeBytes() {
             SystemConfig.METRICSMGR_NETWORK_OPTIONS_SOCKET_SEND_BUFFER_SIZE_BYTES));
   }
 
+  public int getHeronMetricsMaxExceptionsCount() {
+    Object value = this.config.get(
+        SystemConfig.HERON_METRICS_MAX_EXCEPTIONS_COUNT);
+    return value == null ? Integer.MAX_VALUE : TypeUtils.getInteger(value);
```
Comment (billonahill) [inline heron/common/src/java/com/twitter/heron/common/config/SystemConfig.java]: This pattern is repeated in this class. Could we create a methods like this in this class to simplify this:
```
public static int getValue(Map<String, Object> config, String key, Integer default);
```
might be better to do in another pr.

File: heron/common/src/java/com/twitter/heron/common/config/SystemConfig.java [hunk 2]
```diff
@@ -59,6 +59,12 @@
       = "heron.metrics.export.interval.sec";
 
   /**
+   * The maximum number of exceptions in one MetricPublisherPublishMessage protobuf
+   */
+  public static final String HERON_METRICS_MAX_EXCEPTIONS_COUNT
+      = "heron.metrics.max.exceptions.count";
```
Comment (billonahill) [inline heron/common/src/java/com/twitter/heron/common/config/SystemConfig.java]: This name on it's own doesn't clearly convey it's meaning. Can we change it to convey that it's the max exceptions in a protobuf? Something like `heron.metrics.max.exceptions.per.message.count` or something else more clear.

File: heron/common/src/java/com/twitter/heron/common/utils/logging/ErrorReportLoggingHandler.java [hunk 3]
```diff
@@ -75,13 +85,20 @@ public void publish(LogRecord record) {
     // Convert Log
     Throwable throwable = record.getThrown();
     if (throwable != null) {
-      StringWriter sink = new StringWriter();
-      throwable.printStackTrace(new PrintWriter(sink, true));
-      String trace = sink.toString();
       synchronized (ExceptionRepositoryAsMetrics.INSTANCE) {
+        // We would not include the message if already exceeded the exceptions limit
+        if (ExceptionRepositoryAsMetrics.INSTANCE.getExceptionsCount() >= exceptionsLimit) {
```
Comment (billonahill) [inline heron/common/src/java/com/twitter/heron/common/utils/logging/ErrorReportLoggingHandler.java]: What's the scope of this counter? Does this mean that no more exceptions will ever be published, or just for this payload? Same for `droppedExceptionsCount`. Is that count per message payload, count since the jvm started, or what?

Comment (maosongfu) [inline heron/common/src/java/com/twitter/heron/common/utils/logging/ErrorReportLoggingHandler.java]: Both this counter and `droppedExceptionsCount ` are per message payload. 

Every time we collect the metrics, we would invoke IMetric.getAndReset(), send the message out and these two counter will reset to 0.

File: heron/common/src/java/com/twitter/heron/common/config/SystemConfig.java (hunk 1 above)
Comment (maosongfu) [inline heron/common/src/java/com/twitter/heron/common/config/SystemConfig.java]: will do it in another pr

**Comment by billonahill** [comment]
👍
